File: apps/webui/src/lib/api/jobs.ts
import type {
  BuildArtifact,
  BuildJobListResponse,
  BuildJobResponse,
  JobArtifactListResponse,
  JobArtifactMetaResponse,
  JobResourceUsageListResponse,
  JobResourceUsageResponse,
  PruneJobsResponse,
  TimeRange,
  TimeSeriesResponse,
} from "../types";
import { downloadStream, request } from "./client";

interface ListJobsOptions {
  limit?: number;
  offset?: number;
  status?: string;
  packageName?: string;
  mockChroot?: string;
  includeDeleted?: boolean;
}

function listJobsByScope(
  scope: "active" | "completed",
  options: ListJobsOptions = {},
): Promise<BuildJobListResponse> {
  const params = new URLSearchParams();
  params.set("scope", scope);
  if (options.limit !== undefined) params.set("limit", String(options.limit));
  if (options.offset !== undefined) params.set("offset", String(options.offset));
  if (options.status && options.status !== "all") {
    params.set("status", options.status);
  }
  if (options.packageName?.trim()) {
    params.set("package_name", options.packageName.trim());
  }
  if (options.mockChroot?.trim()) {
    params.set("mock_chroot", options.mockChroot.trim());
  }
  if (options.includeDeleted) {
    params.set("include_deleted", "true");
  }
  return request("GET", `/api/v1/jobs?${params.toString()}`);
}

export function listCompletedJobs(
  options: ListJobsOptions = {},
): Promise<BuildJobListResponse> {
  return listJobsByScope("completed", options);
}

export function listActiveJobs(
  options: Omit<ListJobsOptions, "status"> = {},
): Promise<BuildJobListResponse> {
  return listJobsByScope("active", options);
}

export function getJob(id: string): Promise<BuildJobResponse> {
  return request("GET", `/api/v1/jobs/${encodeURIComponent(id)}`);
}

export function listJobUsage(): Promise<JobResourceUsageListResponse> {
  // Live usage feed, not paginated UI; request the server max so no
  // active job's sample is truncated.
  return request("GET", "/api/v1/jobs/usage?limit=200&offset=0");
}

export function getJobUsage(id: string): Promise<JobResourceUsageResponse> {
  return request("GET", `/api/v1/jobs/${encodeURIComponent(id)}/usage`);
}

export function listJobArtifacts(
  id: string,
  limit = 50,
  offset = 0,
): Promise<JobArtifactListResponse> {
  const params = new URLSearchParams({
    limit: String(limit),
    offset: String(offset),
  });
  return request(
    "GET",
    `/api/v1/jobs/${encodeURIComponent(id)}/artifacts?${params.toString()}`,
  );
}

export function getJobArtifactMeta(
  id: string,
  file: string,
): Promise<JobArtifactMetaResponse> {
  return request(
    "GET",
    `/api/v1/jobs/${encodeURIComponent(id)}/artifacts/${file
      .split("/")
      .map(encodeURIComponent)
      .join("/")}/meta`,
  );
}

/**
 * Saves already-buffered log text (collected from the SSE stream) to a file.
 * The chunk/meta byte-range endpoints were removed in favour of the streaming
 * log API, so downloads are produced client-side from the in-memory buffer.
 */
export function downloadLogText(text: string, source: string): void {
  const blob = new Blob([text], { type: "text/plain;charset=utf-8" });
  const filename = source.split("/").pop() || source;
  triggerDownload(blob, filename);
}

export function deleteJob(id: string): Promise<void> {
  return request<void>("DELETE", `/api/v1/jobs/${encodeURIComponent(id)}`);
}

export function killJob(id: string): Promise<BuildJobResponse> {
  return request("POST", `/api/v1/jobs/${encodeURIComponent(id)}/kill`, {});
}

export function retryJob(id: string): Promise<BuildJobResponse> {
  return request("POST", `/api/v1/jobs/${encodeURIComponent(id)}/retry`, {});
}

export function pruneFailedJobs(): Promise<PruneJobsResponse> {
  return request("POST", "/api/v1/jobs/prune-failed", {});
}

export function getJobsTimeseries(
  range: TimeRange = "24h",
): Promise<TimeSeriesResponse> {
  return request("GET", `/api/v1/jobs/timeseries?range=${range}`);
}

export async function downloadJobArtifact(
  id: string,
  artifact: BuildArtifact,
): Promise<void> {
  const path = `/api/v1/jobs/${encodeURIComponent(id)}/artifacts/${artifact.file
    .split("/")
    .map((segment) => encodeURIComponent(segment))
    .join("/")}/content`;

  const res = await downloadStream(path);
  const blob = await res.blob();
  const disposition = res.headers.get("content-disposition") || "";
  const match = disposition.match(/filename="([^"]+)"/i);
  const fallbackName = artifact.file.split("/").at(-1) || "artifact.bin";
  triggerDownload(blob, match?.[1] || fallbackName);
}

function triggerDownload(blob: Blob, filename: string): void {
  const objectUrl = window.URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = objectUrl;
  anchor.download = filename;
  document.body.appendChild(anchor);
  anchor.click();
  anchor.remove();
  window.URL.revokeObjectURL(objectUrl);
}

File: apps/rust/daemon/src/http/jobs.rs
use super::{AppError, AppState};
use axum::Json;
use axum::Router;
use axum::extract::{Path, Query, State};
use axum::routing::{get, post};
use synforge_core::api::{
    BuildJobListResponse, BuildJobResponse, JobListQuery, JobListScope,
    JobResourceUsageListResponse, JobResourceUsageResponse, PruneJobsResponse, TimeSeriesQuery,
    TimeSeriesResponse,
};
use uuid::Uuid;

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/jobs", get(list_jobs))
        .route("/jobs/usage", get(list_job_usage))
        .route("/jobs/timeseries", get(get_jobs_timeseries))
        .route("/jobs/prune-failed", post(prune_failed_jobs))
        .route("/jobs/{id}/kill", post(kill_job))
        .route("/jobs/{id}/retry", post(retry_job))
        .route("/jobs/{id}", get(get_job).delete(delete_job))
        .route("/jobs/{id}/usage", get(get_job_usage))
}

#[utoipa::path(
    get,
    path = "/api/v1/jobs/timeseries",
    tag = "Jobs",
    params(TimeSeriesQuery),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Bucketed completed-job counts over time", body = TimeSeriesResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn get_jobs_timeseries(
    State(state): State<AppState>,
    Query(query): Query<TimeSeriesQuery>,
) -> Result<Json<TimeSeriesResponse>, AppError> {
    Ok(Json(state.service.get_jobs_timeseries(query.range).await?))
}

#[utoipa::path(
    get,
    path = "/api/v1/jobs",
    tag = "Jobs",
    params(JobListQuery),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "List all jobs", body = BuildJobListResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn list_jobs(
    State(state): State<AppState>,
    Query(query): Query<JobListQuery>,
) -> Result<Json<BuildJobListResponse>, AppError> {
    let scope = query.scope.unwrap_or(JobListScope::All);
    let include_deleted = query.include_deleted.unwrap_or(false);
    let response = match scope {
        JobListScope::All => {
            state
                .service
                .list_jobs(
                    query.limit,
                    query.offset,
                    query.status,
                    query.package_name,
                    query.mock_chroot,
                    include_deleted,
                )
                .await?
        }
        JobListScope::Active => {
            state
                .service
                .list_active_jobs(
                    query.limit,
                    query.offset,
                    query.package_name,
                    query.mock_chroot,
                )
                .await?
        }
        JobListScope::Completed => {
            state
                .service
                .list_completed_jobs(
                    query.limit,
                    query.offset,
                    query.status,
                    query.package_name,
                    query.mock_chroot,
                    include_deleted,
                )
                .await?
        }
    };
    Ok(Json(response))
}

#[utoipa::path(
    get,
    path = "/api/v1/jobs/{id}",
    tag = "Jobs",
    params(
        ("id" = Uuid, Path, description = "Job identifier")
    ),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Get job", body = BuildJobResponse),
        (status = 401, body = synforge_core::api::ApiError),
        (status = 404, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn get_job(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<BuildJobResponse>, AppError> {
    Ok(Json(state.service.get_job(id).await?))
}

#[utoipa::path(
    get,
    path = "/api/v1/jobs/usage",
    tag = "Jobs",
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "List live usage for active jobs", body = JobResourceUsageListResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn list_job_usage(
    State(state): State<AppState>,
) -> Result<Json<JobResourceUsageListResponse>, AppError> {
    Ok(Json(state.service.list_job_resource_usage().await?))
}

#[utoipa::path(
    get,
    path = "/api/v1/jobs/{id}/usage",
    tag = "Jobs",
    params(
        ("id" = Uuid, Path, description = "Job identifier")
    ),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Get live usage for one job", body = JobResourceUsageResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn get_job_usage(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<JobResourceUsageResponse>, AppError> {
    Ok(Json(state.service.get_job_resource_usage(id).await?))
}

#[utoipa::path(
    post,
    path = "/api/v1/jobs/{id}/kill",
    tag = "Jobs",
    params(
        ("id" = Uuid, Path, description = "Job identifier")
    ),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Kill active job", body = BuildJobResponse),
        (status = 401, body = synforge_core::api::ApiError),
        (status = 404, body = synforge_core::api::ApiError),
        (status = 409, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn kill_job(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<BuildJobResponse>, AppError> {
    Ok(Json(state.service.kill_job(id).await?))
}

#[utoipa::path(
    post,
    path = "/api/v1/jobs/{id}/retry",
    tag = "Jobs",
    params(
        ("id" = Uuid, Path, description = "Job identifier")
    ),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Retry a finished job", body = BuildJobResponse),
        (status = 400, body = synforge_core::api::ApiError),
        (status = 401, body = synforge_core::api::ApiError),
        (status = 404, body = synforge_core::api::ApiError),
        (status = 409, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn retry_job(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<BuildJobResponse>, AppError> {
    Ok(Json(state.service.retry_job(id).await?))
}

#[utoipa::path(
    delete,
    path = "/api/v1/jobs/{id}",
    tag = "Jobs",
    params(
        ("id" = Uuid, Path, description = "Job identifier")
    ),
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Delete job", body = BuildJobResponse),
        (status = 401, body = synforge_core::api::ApiError),
        (status = 404, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn delete_job(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> Result<Json<BuildJobResponse>, AppError> {
    Ok(Json(state.service.delete_job(id).await?))
}

#[utoipa::path(
    post,
    path = "/api/v1/jobs/prune-failed",
    tag = "Jobs",
    security(("session_auth" = [])),
    responses(
        (status = 200, description = "Prune failed and timed out jobs", body = PruneJobsResponse),
        (status = 401, body = synforge_core::api::ApiError)
    )
)]
pub(super) async fn prune_failed_jobs(
    State(state): State<AppState>,
) -> Result<Json<PruneJobsResponse>, AppError> {
    Ok(Json(state.service.prune_failed_jobs().await?))
}
